Remove IDisable

reaperrr opened the pull request:
This removes the remaining uses of `Actor.IsDisabled` and `IDisable`.

Closes #7807.
Closes #10237.
Closes #12480.
Closes #14177.

reaperrr committed:
Make Gate more independent from Building and pausable-conditional
Replace Gate IsDisabled checks with IsTraitDisabled/Paused checks

reaperrr committed:
Remove unit.IsDisabled check from AirStates

reaperrr committed:
Replace IsDisabled checks in production with IsTraitPaused/Disabled checks

Note: We might want to separate IsTraitDisabled checks later (possibly make the latter cancel the currently produced item), but that can be done in a follow-up.

reaperrr committed:
Remove IDisable for good

pchote commented on OpenRA.Mods.Common/Traits/Player/ClassicProductionQueue.cs:
```diff
@@ -104,7 +104,7 @@ protected override bool BuildUnit(ActorInfo unit)
 			}
 
-			foreach (var p in producers.Where(p => !p.Actor.IsDisabled()))
+			foreach (var p in producers.Where(p => !p.Trait.IsTraitPaused && !p.Trait.IsTraitDisabled))
```
I think it would fit better to put the `IsTraitDisabled` check as part of the `producers` lookup above.  If all production traits are disabled (as opposed to paused) then we should cancel the production and give a refund.

pchote commented on OpenRA.Mods.Common/Traits/Player/ProductionQueue.cs:
```diff
@@ -389,7 +389,7 @@ protected virtual bool BuildUnit(ActorInfo unit)
 
 			var sp = self.TraitsImplementing<Production>().FirstOrDefault(p => p.Info.Produces.Contains(Info.Type));
-			if (sp != null && !self.IsDisabled() && sp.Produce(self, unit, inits))
+			if (sp != null && !sp.IsTraitPaused && !sp.IsTraitDisabled && sp.Produce(self, unit, inits))
```
Likewise here.

reaperrr commented:
> OpenRA.Utility(1,1): Error: Missing Type: DisableOnConditionInfo

Where is this hiding?! Can't find it anywhere...

reaperrr commented:
Updated.

penev92 commented:
The dependency was merged.

penev92 commented on OpenRA.Mods.Common/Traits/Buildings/Gate.cs:
```diff
@@ -111,13 +111,12 @@ void INotifyBlockingMove.OnNotifyBlockingMove(Actor self, Actor blocking)
 		{
-			base.AddedToWorld(self);
-			blockedPositions = Info.Tiles(self.Location);
+			blockedPositions = building.Info.Tiles(self.Location);
```
This looks like it can just as safely be done in the constructor because it only uses that trait's `Info`.

penev92 commented on OpenRA.Mods.Common/Traits/Render/WithGateSpriteBody.cs:
```diff
@@ -18,7 +18,7 @@
 namespace OpenRA.Mods.Common.Traits.Render
 {
-	class WithGateSpriteBodyInfo : WithSpriteBodyInfo, IWallConnectorInfo, Requires<GateInfo>
+	class WithGateSpriteBodyInfo : WithSpriteBodyInfo, IWallConnectorInfo, Requires<BuildingInfo>, Requires<GateInfo>
```
I am not against it, but we should be clear about whether we want redundant `requires` (`GateInfo` already requires `BuildingInfo`).

pchote replied:
Relying on transitive dependencies is a massive code smell.  This trait requires `Building` to work correctly, so should declare that dependency independently from what any other trait does.

But, saying that, IMO it would be cleaner to expose the footprint from `Gate` so that this trait doesn't need to reference `Building` at all.

pchote replied:
Relying on transitive dependencies is a massive code smell.  This trait requires `Building` to work correctly, so should declare that dependency independently from what any other trait does.

But, saying that, IMO it would be cleaner to expose the footprint from `Gate` so that this trait doesn't need to reference `Building` at all.

penev92 commented on OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs:
```diff
@@ -1210,6 +1210,77 @@ internal static void UpgradeActorRules(ModData modData, int engineVersion, ref L
+				// Made Gate not inherit Building
+				if (engineVersion < 20171119)
+				{
+					var gate = node.Value.Nodes.FirstOrDefault(n => n.Key.StartsWith("Gate", StringComparison.Ordinal));
```
Any reason to use `StartsWith` ?

reaperrr commented:
Rebased and updated.

pchote commented on OpenRA.Mods.Common/AI/States/AirStates.cs:
```diff
@@ -31,8 +31,8 @@ protected static int CountAntiAirUnits(IEnumerable<Actor> units)
-				if (unit != null && unit.Info.HasTraitInfo<AttackBaseInfo>() && !unit.Info.HasTraitInfo<AircraftInfo>()
-					&& !unit.IsDisabled())
+				if (unit != null && !unit.Info.HasTraitInfo<AircraftInfo>() &&
+					unit.TraitsImplementing<AttackBase>().Any(ab => !ab.IsTraitPaused && !ab.IsTraitDisabled))
```
We can use this opportunity to check against the *correct* armaments, and also reduce the number of queries:
```csharp
if (unit == null || unit.Info.HasTraitInfo<AircraftInfo>())
	continue;

var arms = unit.TraitsImplementing<AttackBase>()
	.Where(ab => !ab.IsTraitPaused && !ab.IsTraitDisabled)
	.SelectMany(ab => ab.Armaments);
```

pchote commented on OpenRA.Mods.Common/Traits/Production.cs:
```diff
@@ -19,27 +19,26 @@
 	}
 
-	public class Production : INotifyCreated
+	public class Production : PausableConditionalTrait<ProductionInfo>, INotifyCreated
```
The `Produce` method below should check and `return false` if paused.

reaperrr replied:
Done.

pchote commented on OpenRA.Mods.Common/Traits/Buildings/Gate.cs:
```diff
@@ -111,13 +112,7 @@ void INotifyBlockingMove.OnNotifyBlockingMove(Actor self, Actor blocking)
-		protected override void AddedToWorld(Actor self)
-		{
-			base.AddedToWorld(self);
-			blockedPositions = Info.Tiles(self.Location);
```
`blockedPositions` still needs to be updated if the actor is manually removed and readded to the world (e.g. using Lua).

pchote reviewed (dismissed):
Looking good! Just a couple of minor code nits that i'd like you to address before I dig into ingame testing:

reaperrr commented:
Updated.

penev92 commented on OpenRA.Mods.Common/Traits/Buildings/Gate.cs:
```diff
@@ -109,15 +112,14 @@ void INotifyBlockingMove.OnNotifyBlockingMove(Actor self, Actor blocking)
 		{
-			base.AddedToWorld(self);
-			blockedPositions = Info.Tiles(self.Location);
+			return !IsTraitDisabled && !IsTraitPaused && building.BuildComplete && blocking.AppearsFriendlyTo(self);
```
This seems to be missing a `&& !building.Locked` check to be consistent with the one in `Tick()`.

penev92 commented on OpenRA.Mods.Common/Traits/Buildings/Gate.cs:
```diff
@@ -109,15 +112,14 @@ void INotifyBlockingMove.OnNotifyBlockingMove(Actor self, Actor blocking)
 		}
 
-		bool CanRemoveBlockage(Actor self, Actor blocking)
+		void INotifyAddedToWorld.AddedToWorld(Actor self)
```
This also needs
```
		void INotifyRemovedFromWorld.RemovedFromWorld(Actor self)
		{
			blockedPositions = Enumerable.Empty<CPos>();
		}
```

to fully address @pchote's earlier comment:

>`blockedPositions` still needs to be updated if the actor is manually removed and readded to the world (e.g. using Lua).

penev92 replied:
Done.

penev92 commented on mods/ts/rules/defaults.yaml:
```diff
@@ -1160,30 +1156,37 @@
 		WallConnections: -1,0, 3,0
 	LineBuildNode:
 		Connections: -1,0, 1,0
+	Gate:
```
`^Gate_A` already inherits `^Gate`, which has the `Gate` trait, so this is redundant.

penev92 commented on mods/ts/rules/defaults.yaml:
```diff
@@ -1160,30 +1156,37 @@
 		WallConnections: 0,-1, 0,3
 	LineBuildNode:
 		Connections: 0,-1, 0,1
+	Gate:
```
`^Gate_B` already inherits `^Gate`, which has the `Gate` trait, so this is redundant.

reaperrr replied:
For the record: those redundancies came from the upgrade rule, I just didn't check for redundancies (due to lazyness, mostly).

penev92 commented on mods/ts/rules/defaults.yaml:
```diff
@@ -1144,12 +1143,9 @@
 		OpenSequence: open
 	Tooltip:
-	Gate:
+	Building:
```
We have `-Building:` just a few lines above? Something smells fishy here.

penev92 replied:
Done.

penev92 reviewed (dismissed):
Leaving a bunch of change requests. Some of them may need to be discussed first, like the LINQ vs `foreach` one.

penev92 commented:
Also here are some comments that I couldn't fit in the review because of GitHub being limiting about where you can put comments in the diffs:

 - `ProductionQueue.MostLikelyProducer` seems like it should start respecting the fact that `Production` is now `Pausable`/`Disablable`.
 - Same goes for `ClassicProductionQueue`'s `Tick`, `MostLikelyProducer` and `GetBuildTime`, since they all have references to `Production`.
 - I have not checked the RA rule definitions for the gates, but the TS ones needed some fixing, so I'm going to assume those do as well.

pchote commented on OpenRA.Mods.Common/AI/States/AirStates.cs:
```diff
@@ -31,17 +31,19 @@ protected static int CountAntiAirUnits(IEnumerable<Actor> units)
+					.Where(ab => !ab.IsTraitPaused && !ab.IsTraitDisabled)
+					.SelectMany(ab => ab.Armaments);
+
+				foreach (var a in arms)
```
AttackBase does not necessarily control all the armaments on an actor.  Selecting all armaments ignoring AttackBase would be wrong.

penev92 replied:
Meaning we can have armaments that are not used by any attacking trait?

reaperrr replied:
Re: performance/foreach loops: In general I agree that the AI code is probably one of the lower-hanging fruits in terms of performance-wins, but since this PR is primarily about removing IDisable, I think that should be left for later (or a separate PR, in any case).

reaperrr commented on OpenRA.Mods.Common/Traits/Player/ProductionQueue.cs:
```diff
@@ -375,8 +375,10 @@ public virtual TraitPair<Production> MostLikelyProducer()
 		{
-			// Cannot produce if I'm dead
-			if (!self.IsInWorld || self.IsDead)
+			var productionTraits = self.TraitsImplementing<Production>().Where(Exts.IsTraitEnabled);
```
Then we'd need ro `Requires` Production, which might cause circular dependencies.
I'll use INotifyCreated instead, that's safer.

reaperrr commented:
Updated.

penev92 commented on OpenRA.Mods.Common/Traits/Player/ClassicProductionQueue.cs:
```diff
@@ -54,6 +54,9 @@ protected override void Tick(Actor self)
 			isActive = false;
 			foreach (var x in self.World.ActorsWithTrait<Production>())
 			{
+				if (x.Trait.IsTraitDisabled)
```
I now wonder if `ProductionQueue.Tick()` also needs such a check.

reaperrr replied:
I now made it so that `ProductionQueue.Tick` cancels current production incl. queued items completely when all production traits are disabled, and at least doesn't tick the queue when they are paused. Hope that works as intended...

penev92 commented:
Code looks mostly good (nothing was changed to the `AirStates` code), with a few notes:
 - `ProductionQueue.Tick` may or may not want to check `IsTraitDisabled`/`IsTraitPaused`.
 - `ClassicProductionQueue.GetBuildTime` more than likely wants to check those.
 - Those two lead to the question "does `ProductionQueue.GetBuildTime` *also* want to check that?"
 - I haven't checked RA gate rule definitions yet, but the TS rules look good now.
 - Have not yet tested anything ingame.

penev92 commented on OpenRA.Mods.Common/Traits/Production.cs:
```diff
@@ -122,7 +121,7 @@ public virtual void DoProduction(Actor self, ActorInfo producee, ExitInfo exitin
 		public virtual bool Produce(Actor self, ActorInfo producee, TypeDictionary inits)
 		{
-			if (Reservable.IsReserved(self) || (building != null && building.Locked))
+			if (IsTraitPaused || Reservable.IsReserved(self) || (building != null && building.Locked))
```
Done.

pchote replied:
As above, it would be a good idea to check for `IsTraitDisabled` here too, just in case.

reaperrr commented:
> Those two lead to the question "does ProductionQueue.GetBuildTime also want to check that?"

In my opinion no, because it only calculates the time for that item once before its production would be started ( you can tell by the fact that cheats or low power have no impact on build speed of actors already in production).

reaperrr commented:
Updated.

penev92 commented on OpenRA.Mods.Common/Traits/Player/ProductionQueue.cs:
```diff
@@ -237,13 +245,19 @@ void ITick.Tick(Actor self)
 
 		protected virtual void Tick(Actor self)
 		{
+			if (queue.Count > 0 && enabledProductionTraits.Count == 0)
```
I somehow managed to get an NRE here with `queue.Count` being 1 and `enabledProductionTraits` being `null`, but `productionTraits` being an empty collection.
You should initialize `enabledProductionTraits` in `INotifyCreated.Created` as well to avoid that, and probably change this to `queue.Count > 0 && productionTraits.Count > 0 && enabledProductionTraits.Count == 0` to skip some queue clearing when `productionTraits` is empty (although I'm not sure how that's empty).
Also `ClearQueue` contains a `queue.Count` check, so that's redundant.

reaperrr replied:
> when productionTraits is empty (although I'm not sure how that's empty).

Could that be because it's a list and not an array or enumerable?

reaperrr commented:
Updated, but the `productionTraits`/`enabledProductionTraits` usage may need another closer look.

pchote commented on OpenRA.Mods.Cnc/Traits/Buildings/ProductionAirdrop.cs:
```diff
@@ -41,6 +41,9 @@ public ProductionAirdrop(ActorInitializer init, ProductionAirdropInfo info)
 
 		public override bool Produce(Actor self, ActorInfo producee, TypeDictionary inits)
 		{
+			if (IsTraitPaused)
```
To be safe, can we please also check `IsTraitDisabled` here?

pchote commented on mods/ts/rules/defaults.yaml:
```diff
@@ -1136,20 +1137,16 @@
 	Tooltip:
-	Gate:
+	Building:
 		BuildSounds: place2.aud
```
This is inherited from `^BasicBuilding`, so can be manually removed.

pchote commented on mods/ts/rules/defaults.yaml:
```diff
@@ -1136,20 +1137,16 @@
 		BuildSounds: place2.aud
-		OpeningSound: gateup1.aud
-		ClosingSound: gatedwn1.aud
 		TerrainTypes: Clear, Rough, Road, DirtRoad, Green, Sand, Pavement
```
...But note that this one *isn't* inherited (it matches walls, which can be placed on more terrain types – is this correct?)

reaperrr replied:
I'm 100% sure Green, Sand and Pavement were valid terrain for *all* buildings in TS.

The only potentially wrong terrain type here is 'Rough', but ^BasicBuilding has that as well. I'll just move this list to ^BasicBuilding for now.

pchote commented on OpenRA.Mods.Common/Traits/Player/ClassicProductionQueue.cs:
```diff
@@ -104,7 +108,7 @@ protected override bool BuildUnit(ActorInfo unit)
 			}
 
-			foreach (var p in producers.Where(p => !p.Actor.IsDisabled()))
+			foreach (var p in producers.Where(p => !p.Trait.IsTraitPaused))
```
Minor perf nit: no benefit to using LINQ here, this can move to an `if () continue` at the top of the loop.

pchote commented on OpenRA.Mods.Common/Traits/Player/ProductionQueue.cs:
```diff
@@ -76,6 +76,9 @@ public class ProductionQueue : IResolveOrder, ITick, ITechTreeElement, INotifyOw
 		readonly IEnumerable<ActorInfo> allProducibles;
 		readonly IEnumerable<ActorInfo> buildableProducibles;
 
+		List<Production> productionTraits;
```
This will never change, so better an array than a List.

pchote commented on OpenRA.Mods.Common/Traits/Player/ProductionQueue.cs:
```diff
@@ -237,13 +246,19 @@ void ITick.Tick(Actor self)
 
 		protected virtual void Tick(Actor self)
 		{
+			if (productionTraits.Count > 0 && enabledProductionTraits.Count == 0)
```
What is the purpose of `productionTraits.Count > 0` here?  `ProductionQueue` requires `Production` traits to live on the same actor, so won't work at all if there are none.  Immediately clearing the queue in this case seems reasonable to me.

pchote commented on OpenRA.Mods.Common/Traits/Player/ProductionQueue.cs:
```diff
@@ -76,6 +76,9 @@ public class ProductionQueue : IResolveOrder, ITick, ITechTreeElement, INotifyOw
 		readonly IEnumerable<ActorInfo> buildableProducibles;
 
+		List<Production> productionTraits;
+		List<Production> enabledProductionTraits;
```
I don't like the look how how this is used.  It would be cleaner and more conventional to use `productionTraits.Where(not disabled)` instead.

pchote commented on OpenRA.Mods.Common/Traits/ProductionFromMapEdge.cs:
```diff
@@ -43,6 +43,9 @@ void INotifyCreated.Created(Actor self)
 
 		public override bool Produce(Actor self, ActorInfo producee, TypeDictionary inits)
 		{
+			if (IsTraitPaused)
```
And here.

pchote commented on OpenRA.Mods.Common/Traits/ProductionParadrop.cs:
```diff
@@ -46,6 +46,9 @@ public ProductionParadrop(ActorInitializer init, ProductionParadropInfo info)
 
 		public override bool Produce(Actor self, ActorInfo producee, TypeDictionary inits)
 		{
+			if (IsTraitPaused)
```
And here.

reaperrr commented:
Updated and rebased,

reaperrr committed:
Replace LINQ in AirStates.CountAntiAirUnits

reaperrr commented:
Rebased and updated.
- Fixed broken ClassicProductionQueue
- Fixed that even after above fix, ClassicProductionQueue would continue ticking when it should be paused
- Avoid LINQ in AirStates

GraionDilach commented:
Needs another rebase now.

penev92 commented on OpenRA.Mods.Common/Traits/Player/ProductionQueue.cs:
```diff
@@ -236,6 +243,26 @@ void ITick.Tick(Actor self)
+			QueueTick(self, !anyUnpausedProduction);
+		}
+
+		protected virtual void QueueTick(Actor self, bool allProductionPaused)
```
Can we go with `TickQueue` please.

penev92 commented on OpenRA.Mods.Common/Traits/Player/ProductionQueue.cs:
```diff
@@ -236,6 +243,26 @@ void ITick.Tick(Actor self)
+			QueueTick(self, !anyUnpausedProduction);
+		}
+
+		protected virtual void QueueTick(Actor self, bool allProductionPaused)
```
Can we go with `TickQueue` please.

pchote replied:
`TickInner` would be more in line with other parts of the codebase.

penev92 commented on OpenRA.Mods.Common/Traits/Player/ClassicProductionQueue.cs:
```diff
@@ -52,16 +52,23 @@ protected override void Tick(Actor self)
 		{
 			// PERF: Avoid LINQ.
 			isActive = false;
+			var isActiveUnpaused = false;
```
I don't understand this.

reaperrr replied:
I couldn't come up with anything better that isn't overly long.

Basically, this checks whether the queue is neither disabled nor paused.

pchote replied:
Why can't we track that using `isActive`?

reaperrr replied:
Wouldn't that make IsTraitDisabled and IsTraitPaused have the same effect?

I'm wary of that since isActive is used in other places.

Besides, I think the most correct behavior would be for the queue to be cleared when all Productions are disabled, but just not tick when some are enabled but all of those enabled are paused. That's not possible if we use isActive alone.

pchote replied:
In that case IMO the real problem is with the `isActive` name.  `isActive` should be used for pausing, while `isEnabled` or similar should be used for disabled/clearing production.

penev92 reviewed (commented):
Disabling all production structures for a production tab has a pretty crap interaction with the UI, but I'm not sure what we want to do there, and is probably not for this PR.
Also EMP-ing a ConYard in TS doesn't pause production. Is that intentional?

pchote commented on OpenRA.Mods.Common/Traits/Player/ProductionQueue.cs:
```diff
@@ -244,8 +271,10 @@ protected virtual void Tick(Actor self)
-			if (queue.Count > 0)
-				queue[0].Tick(playerResources);
+			if (queue.Count < 1 || allProductionPaused)
+				return;
```
Changing this to use an early return feels unnecessary.  Can you please stick with `queue.Count > 0 && !allProductionPaused`?

reaperrr commented:
Rebased and updated.

> Disabling all production structures for a production tab has a pretty crap interaction with the UI, but I'm not sure what we want to do there, and is probably not for this PR.

Agree on both accounts (unless it's easy to fix and someone tells me how).

penev92 commented on OpenRA.Mods.Common/Traits/Player/ProductionQueue.cs:
```diff
@@ -89,9 +91,10 @@ public class ProductionQueue : IResolveOrder, ITick, ITechTreeElement, INotifyOw
+		[Sync] public bool Enabled { get; protected set; }
 
 		public string Faction { get; private set; }
+		[Sync] public bool ValidFaction { get; private set; }
```
I don't understand why you folks so dislike the "Is" prefix on boolean properties :/

penev92 reviewed (dismissed):
👍

pchote commented on OpenRA.Mods.Common/Traits/Player/ProductionQueue.cs:
```diff
@@ -102,14 +105,20 @@ public ProductionQueue(ActorInitializer init, Actor playerActor, ProductionQueue
 			Faction = init.Contains<FactionInit>() ? init.Get<FactionInit, string>() : self.Owner.Faction.InternalName;
-			Enabled = !info.Factions.Any() || info.Factions.Contains(Faction);
+			ValidFaction = !info.Factions.Any() || info.Factions.Contains(Faction);
+			Enabled = ValidFaction;
```
In a future PR we could consider replacing the special case faction logic with a faction-granted condition.

pchote commented on OpenRA.Mods.Common/Traits/Player/ProductionQueue.cs:
```diff
@@ -367,16 +398,18 @@ protected void BeginProduction(ProductionItem item)
 		{
-			// Cannot produce if I'm dead
-			if (!self.IsInWorld || self.IsDead)
+			var enabledProductionTraits = productionTraits.Where(p => !p.IsTraitDisabled).ToArray();
```
The uses of this below usually won't need to enumerate the whole collection, so the `ToArray()` will actually hurt performance.

pchote replied:
The best fix is to query `MostLikelyProducer` hereish, cancel the production if the trait is null, and then return false if it is paused or disabled.

pchote commented on OpenRA.Mods.Common/Traits/Player/ProductionQueue.cs:
```diff
@@ -236,6 +245,28 @@ void ITick.Tick(Actor self)
+			var anyUnpausedProduction = false;
+			foreach (var p in productionTraits)
+			{
+				if (!anyEnabledProduction)
```
These checks aren't needed, it is more efficient to use `anyEnabledProduction |= !p.IsTraitDisabled;` and so on.  Can you please change this in all the places you use this pattern?

reaperrr replied:
Apart from the two here, I'm not aware of any other cases?

pchote commented on OpenRA.Mods.Common/Traits/Player/ClassicProductionQueue.cs:
```diff
@@ -46,39 +46,47 @@ public ClassicProductionQueue(ActorInitializer init, ClassicProductionQueueInfo
+					&& !x.Trait.IsTraitDisabled && !x.Trait.IsTraitPaused && x.Trait.Info.Produces.Contains(Info.Type))
 				.OrderByDescending(x => x.Actor.IsPrimaryBuilding())
 				.ThenByDescending(x => x.Actor.ActorID)
 				.FirstOrDefault();
```
Please change this to
```csharp
			var productionActors = self.World.ActorsWithTrait<Production>()
				.Where(x => x.Actor.Owner == self.Owner
					&& !x.Trait.IsTraitDisabled && x.Trait.Info.Produces.Contains(Info.Type))
				.OrderByDescending(x => x.Actor.IsPrimaryBuilding())
				.ThenByDescending(x => x.Actor.ActorID)
				.ToList();

			var unpaused = productionActors.FirstOrDefault(a => !a.Trait.IsTraitPaused);
			return unpaused.Trait != null ? unpaused : productionActors.FirstOrDefault();
```

to fix the sidebar glitches when a production structure in TS is EMPed.

pchote commented on OpenRA.Mods.Common/Traits/Player/ProductionQueue.cs:
```diff
@@ -367,16 +398,18 @@ protected void BeginProduction(ProductionItem item)
 		public virtual TraitPair<Production> MostLikelyProducer()
 		{
-			var trait = self.TraitsImplementing<Production>().FirstOrDefault(p => p.Info.Produces.Contains(Info.Type));
+			var trait = productionTraits.FirstOrDefault(p => !p.IsTraitDisabled && !p.IsTraitPaused && p.Info.Produces.Contains(Info.Type));
```
Similarly here: check if the first unpaused one is null and fall back to the first paused one.

pchote reviewed (commented):
I'm sorry... I want to get this merged as much as you do, but:

pchote commented on OpenRA.Mods.Common/Traits/Player/ClassicProductionQueue.cs:
```diff
@@ -46,39 +46,47 @@ public ClassicProductionQueue(ActorInitializer init, ClassicProductionQueueInfo
+				if (x.Actor.Owner != self.Owner || !x.Trait.Info.Produces.Contains(Info.Type))
+					continue;
+
+				if (!Enabled)
```
Two more here

reaperrr commented:
Updated. I hope this is it...

pchote reviewed (approved):
There are a couple of potential nits, but it is not worth delaying this further on them.  Any further issues can be dealt with in followup PRs.

pchote commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/9491d48738c337590a71a6b87f38110f56e74e37
